Add support for serializing the textual representation of LLVM IR.

phoebesv opened the pull request:
This PR adds support for direct serialization of LLVM IR textual representation from the module, eliminating the need for developers to use the llvm-dis tool to convert bitcode into human-readable IR format. This enhancement simplifies debugging and optimization processes, providing developers with a more streamlined workflow.

benvanik reviewed (changes requested):
Useful! LLVM IR is not a universal concept, though, so please remove the new flag and continue to use the old one (as it looks like this anyway?).

You should also add this to the other targets emitting LLVMIR (LLVMGPU and ROCM today) for consistency.

ScottTodd commented on compiler/plugins/target/LLVMCPU/LLVMCPUTarget.cpp:
```diff
@@ -467,6 +477,8 @@ class LLVMCPUTargetBackend final : public TargetBackend {
+      dumpLLVMIRToPath(options.dumpIntermediatesPath, options.dumpBaseName,
+                       variantOp.getName(), ".codegen.ll", *llvmModule);
       dumpBitcodeToPath(options.dumpIntermediatesPath, options.dumpBaseName,
                         variantOp.getName(), ".codegen.bc", *llvmModule);
```
We could update the docs here after this change: https://iree.dev/developers/general/developer-tips/#__tabbed_1_1

* include `.ll` files in sample output
* clarify that `llvm-dis` on the `.bc` files isn't needed when the `.ll` files already exist

![image](https://github.com/iree-org/iree/assets/4010439/517ff4d9-625b-4c20-8139-298ea0773f1e)

Source for that is at https://github.com/iree-org/iree/blob/main/docs/website/docs/developers/general/developer-tips.md

ScottTodd commented on compiler/src/iree/compiler/Dialect/HAL/Target/TargetOptions.cpp:
```diff
@@ -71,11 +73,17 @@ void TargetOptions::bindOptions(OptionsBinder &binder) {
+      "iree-hal-dump-llvm-ir-to", executableLLVMIRPath,
+      llvm::cl::desc(
+          "Path to write translated and serialized llvmir (.ll) into."),
+      llvm::cl::cat(halTargetOptionsCategory));
```
> Useful! LLVM IR is not a universal concept, though, so please remove the new flag and continue to use the old one (as it looks like this anyway?).

+1, I'd just bundle these together and output both/all intermediate files when `iree-hal-dump-executable-intermediates-to` is set

That matches what we do for WebGPU, dumping both the text (human readable) and binary (machine readable) files: https://github.com/iree-org/iree/blob/cc0c8266aa4d457705fcbb94804696ee0c8dcf18/compiler/plugins/target/WebGPUSPIRV/WebGPUSPIRVTarget.cpp#L220-L240

phoebesv replied:
@benvanik @ScottTodd Thank you for reviewing. I have updated the commit to bundle both intermediate files together, added this change to CUDATarget, and updated the developer documentation. It seems that ROCM already dumps .ll in the source. As I am not very familiar with other targets, please feel free to let me know if I missed anything else. https://github.com/iree-org/iree/blob/e12e4b4badd38151c67bcf6f32f031cea076c8c6/compiler/plugins/target/ROCM/ROCMTarget.cpp#L474-L486

phoebesv committed:
[LLVMCPU] Add support for serializing the textual representation of LLVM IR.

This change eliminates the need for llvm-dis tool by adding support for
direct serialization of LLVM IR textual representation. This enhancement
simplifies debugging and optimization processes, providing developers
with a more streamlined workflow.

Signed-off-by: Phoebe Chen <phoebe.chen@sifive.com>

phoebesv committed:
[CUDA] Add support for serializing the textual representation of LLVM IR from the module.

This change eliminates the need for llvm-dis tool by adding support for
direct serialization of LLVM IR textual representation. This enhancement
simplifies debugging and optimization processes, providing developers
with a more streamlined workflow.

Signed-off-by: Phoebe Chen <phoebe.chen@sifive.com>

phoebesv committed:
Include .ll files in sample output in developer-tips document.

Signed-off-by: Phoebe Chen <phoebe.chen@sifive.com>

ScottTodd commented on compiler/plugins/target/CUDA/CUDATarget.cpp:
```diff
@@ -242,14 +243,26 @@ static std::string produceGpuImage(const CUDAOptions &options,
+                                 StringRef suffix, StringRef extPrefix,
+                                 llvm::Module &module) {
+  // Dump disassembly to path.
+  llvm::SmallVector<char, 0> textData;
```
(just musing, no action needed)

Ah, the existing code had this already, but we usually recommend leaving the `, 0>` off of `llvm::SmallVector` (or using a different data structure depending on the expected size of data). https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h

> In the absence of a well-motivated choice for the number of inlined elements N, it is recommended to use SmallVector<T> (that is, omitting the N). This will choose a default number of inlined elements reasonable for allocation on the stack (for example, trying to keep sizeof(SmallVector<T>) around 64 bytes).

In this case we can expect the data to be maybe several kilobytes? So I'm not sure if we get much value from SmallVector, especially with the inlined size set explicitly to 0...

phoebesv replied:
Here is a discussion about N=0 on the LLVM group (https://groups.google.com/g/llvm-dev/c/q1OyHZy8KVc/m/MIRhwrHeBAAJ). It seems there are some examples of using SmallVector<N, 0> for handling large-size data. However, I follow the documentation in omitting N.

ScottTodd commented on compiler/plugins/target/LLVMCPU/LLVMCPUTarget.cpp:
```diff
@@ -53,14 +54,26 @@ namespace mlir::iree_compiler::IREE::HAL {
-  llvm::WriteBitcodeToFile(module, ostream);
-  dumpDataToPath(path, baseName, suffix, extension,
-                 StringRef(data.data(), data.size()));
+static void dumpLLVMModuleToPath(StringRef path, StringRef baseName,
+                                 StringRef suffix, StringRef extPrefix,
+                                 llvm::Module &module) {
```
If this is duplicated between multiple files, we could move it to https://github.com/iree-org/iree/tree/main/compiler/src/iree/compiler/Dialect/HAL/Utils. @benvanik WDYT?

phoebesv replied:
I've also noticed that the smiliar code snippet has been used on various platforms. We could refactor it in the following pull request if there are no problems moving the similar functions to the path @ScottTodd provided.

ScottTodd reviewed (commented):
This generally LGTM. Would like to hear @benvanik 's opinions too though. My comments could also be addressed in a follow-up (by one of us, or you).

kuhar commented on compiler/plugins/target/CUDA/CUDATarget.cpp:
```diff
@@ -242,14 +243,26 @@ static std::string produceGpuImage(const CUDAOptions &options,
+  llvm::raw_svector_ostream textOstream(textData);
+
+  module.print(textOstream, nullptr);
+  llvm::Twine textExtension = llvm::Twine(extPrefix) + ".ll";
```
We can't use twines for local variables: https://llvm.org/docs/ProgrammersManual.html#llvm-adt-twine-h

Either pass it as a function parameter or materialize as a `std::string`

phoebesv replied:
Updated, thank you for catching that.

kuhar commented on compiler/plugins/target/CUDA/CUDATarget.cpp:
```diff
@@ -242,14 +243,26 @@ static std::string produceGpuImage(const CUDAOptions &options,
+  llvm::raw_svector_ostream binaryOstream(binaryData);
+  // Write the specified module to the specified output stream.
+  llvm::WriteBitcodeToFile(module, binaryOstream);
+  llvm::Twine binaryExtension = llvm::Twine(extPrefix) + ".bc";
```
also here

kuhar commented on compiler/plugins/target/LLVMCPU/LLVMCPUTarget.cpp:
```diff
@@ -53,14 +54,26 @@ namespace mlir::iree_compiler::IREE::HAL {
+  llvm::raw_svector_ostream textOstream(textData);
+
+  module.print(textOstream, nullptr);
+  llvm::Twine textExtension = llvm::Twine(extPrefix) + ".ll";
```
also here

kuhar commented on compiler/plugins/target/LLVMCPU/LLVMCPUTarget.cpp:
```diff
@@ -53,14 +54,26 @@ namespace mlir::iree_compiler::IREE::HAL {
+  llvm::raw_svector_ostream binaryOstream(binaryData);
+  // Write the specified module to the specified output stream.
+  llvm::WriteBitcodeToFile(module, binaryOstream);
+  llvm::Twine binaryExtension = llvm::Twine(extPrefix) + ".bc";
```
also here

ScottTodd reviewed (approved):
LGTM. @benvanik or @kuhar did either of you have other feedback (both requested changes)
